rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,262 @@
+
+To obtain the screenshot mode context data and provide it as a service to
+plugins, this RFC recommends using a URL query string variable with raw data,
+and provide an interface to it in the form of an observable. The Screenshot
+Mode Service plugin will read the data in its `setup` method and provide the
+observable to the data interface in its plugin contract. The interface is
+described below as the `ScreenshotModeServiceApi`.
+
+The new query string parameter could appear in the URL of any Kibana
+application. Adding new URL query string parameters is a non-breaking change,
+and applications would not be impacted by having this additional data in the URL.
+
+Since the service data needs to be an object, the best way to encode it as data
+in a URL is to use Rison encoding. An example of what that would look like in a
+Kibana URL (with linebreaks added):
+
+```
+http://localhost:5601/app/kibana
+?screenshots=(enabled:!t,layout:(height:500,width:400),timezone:PST)
+#/discover?_g=()&_a=(columns:!(_source),index:'6a047550-4e87-11ea-ad93-21278bc67061',interval:auto,query:(language:kuery,query:''),sort:!())
+
+```
+
+That shows an example way of encoding this raw data into the URL as a global
+variable called `screenshots`:
+```
+{
+  "enabled": true,
+  "layout": {
+    "height": 500,
+    "width": 400
+  },
+  "timezone": "PST"
+}
+```
+
+The rest of the URL is needed by the application and is used for recreating a
+certain state of data in the page. In order to be "reportable" these parts of
+the URL must be totally sufficient for recreating everything the user expects
+to see in a report.
+
+The `screenshot` object has to be crafted by a client that is generating a
+request for a screenshot. That can be done by anything, but the implementation
+concerns will be left out to keep the RFC about the details on the service.
+
+The Screenshot Mode Service would be a plugin designed to wrap the raw context data for
+applications to consume, and it is up to an outside feature (i.e.  Kibana
+Reporting) to craft the raw context data. Applications are the parts that make
+this design useful for Kibana, by reading the service data and using it to
+inform its rendering.
+
+As a plugin, the screenshot mode service won't automatically work in the
+background to inject the data into the various running stages of applications.
+This is in line with the pattern on the client-side where all query parameters
+are delegated to the individual applications, letting them do as they wish with
+it. For this RFC, it's an important topic because plugins have to "wire-up"
+this service in their router and UI app, but it will prevent "magical" behavior
+happening.
+
+## Server side
+On the server side, applications will have to interact with the screenshot
+plugin to support custom endpoints include the screenshot data as part of the
+route handling context. Example:
+```
+const screenshotRouter = plugins.screenshots.wrapRouter(router);
```

Comment:
What if we need to implement another mode? Dashboard mode, for example? Should plugin be aware of all of them?

Reply:
I can't think of a reason why not.